src/Http/Routing/src/EndpointRoutingMiddleware.cs

```diff
@@ -53,15 +53,24 @@ internal sealed class EndpointRoutingMiddleware
             _endpointDataSource = new CompositeEndpointDataSource(endpointRouteBuilder.DataSources);
         }
 
-        public async Task Invoke(HttpContext httpContext)
+        public Task Invoke(HttpContext httpContext)
         {
             var feature = new EndpointSelectorContext();
 
             // There's an inherent race condition between waiting for init and accessing the matcher
             // this is OK because once `_matcher` is initialized, it will not be set to null again.
-            var matcher = await InitializeAsync();
+            var matcherTask = InitializeAsync();
+            if (!matcherTask.IsCompletedSuccessfully)
+            {
+                return AwaitMatcher(this, httpContext, feature, matcherTask);
+            }
+
+            var matchTask = matcherTask.Result.MatchAsync(httpContext, feature);
+            if (!matchTask.IsCompletedSuccessfully)
+            {
+                return AwaitMatch(this, httpContext, feature, matchTask);
+            }
 
-            await matcher.MatchAsync(httpContext, feature);
             if (feature.Endpoint != null)
```

Comment:
Can you call CompleteAwaited here to avoid the code duplication? (And give ti a better name?)

Reply:
Sure